File: dynamic_programming/different-ways-to-add-parentheses.cpp
/*
  Given a string expression of numbers and operators, return all possible results from computing all the different possible ways to group numbers and operators. You may return the answer in any order.
  The test cases are generated such that the output values fit in a 32-bit integer and the number of different results does not exceed 104.
  
  Example 1:
    Input: expression = "2-1-1"
    Output: [0,2]
    Explanation:
    ((2-1)-1) = 0 
    (2-(1-1)) = 2
*/
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        if (is_only_digits(expression)) {
            return {stoi(expression)};
        }

        vector<int> results;
        int n = expression.length();

        for (size_t i = 0; i < n; ++i) {
            char c = expression[i];
            if (c == '+' || c == '-' || c == '*') {
                vector<int> left_results = diffWaysToCompute(expression.substr(0, i));
                vector<int> right_results = diffWaysToCompute(expression.substr(i + 1, n));

                for (const int ll : left_results) {
                    for (const int rr : right_results) {
                        int result = 0;
                        switch (c) {
                            case '+':
                                result = ll + rr;
                                break;
                            case '-':
                                result = ll - rr;
                                break;
                            case '*':
                                result = ll * rr;
                                break;
                        }
                        results.push_back(result);
                    }
                }
            }
        }

        return results;
    }
private:
    static bool is_only_digits(const string& expression) {
        for (const char c : expression) {
            if (!isdigit(c)) return false;
        }
        return true;
    }
};

File: dynamic_programming/different-ways-to-add-parentheses.rs
/*
  Given a string expression of numbers and operators, return all possible results from computing all the different possible ways to group numbers and operators. You may return the answer in any order.
  The test cases are generated such that the output values fit in a 32-bit integer and the number of different results does not exceed 104.
  
  Example 1:
    Input: expression = "2-1-1"
    Output: [0,2]
    Explanation:
    ((2-1)-1) = 0 
    (2-(1-1)) = 2
*/
impl Solution {
    pub fn diff_ways_to_compute(expression: String) -> Vec<i32> {
        if expression.chars().all(char::is_numeric) {
            return vec![expression.parse::<i32>().unwrap()];
        }

        let mut results = Vec::new();

        for (i, c) in expression.chars().enumerate() {
            match c {
                '+' | '-' | '*' => {
                    let left_results = Self::diff_ways_to_compute(expression[..i].to_string());
                    let right_results = Self::diff_ways_to_compute(expression[i+1..].to_string());

                    for &left in &left_results {
                        for &right in &right_results {
                            let result = match c {
                                '+' => left + right,
                                '-' => left - right,
                                '*' => left * right,
                                _ => unreachable!(),
                            };
                            results.push(result);
                        }
                    }
                },
                _ => {}
            }
        }

        results
    }
}
